### ISSUE TITLE
Excessive memory consumption while using DistributedDataParallel 

### ISSUE THREAD

**Issue by Sentient07** [pytorch/vision#2390]
## 🐛 Bug

TL;DR : While using `DistributedDataParallel` and multiple GPU, memory consumption on each GPU seems to be more than twice as much as what is observed when without using `DistributedDataParallel` on a single GPU.
 
## To Reproduce

I have been using FasterRCNN from Torchvision’s models that uses DistributedDataParallel for training. However, I find that while using multiple GPU, the memory consumption is far more than without multiple GPU. Here is my code

```
kwargs = {}
  kwargs['min_size'] = args.min_size
  kwargs['max_size'] = args.max_size
  model = ModifiedFRCNN(cfg=cfg, custom_anchor=args.custom_anchor,
                        use_def=args.use_def, cpm=args.cpm,
                        default_filter=args.default_filter,
                        soft_nms=args.soft_nms,
                        upscale_r=args.upscale_r, **kwargs).cuda().eval()
  model = restore_network(model)
  model_without_ddp = model
  dataset = GenData(args.test_dataset,
                    args.base_path,
                    dataset_param=None,
                    train=False)

  if args.n_gpu > 1:
      init_distributed_mode(args)
      model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu],
                                                        find_unused_parameters=True)
      model_without_ddp = model.module
      sampler = torch.utils.data.distributed.DistributedSampler(dataset)
      batch_sampler = torch.utils.data.BatchSampler(sampler,
                                                    args.batch_size,
                                                    drop_last=True)
      data_loader = torch.utils.data.DataLoader(dataset,
                                                batch_sampler=batch_sampler,
                                                num_workers=args.num_workers,
                                                collate_fn=coco_collate)
      metric_logger = MetricLogger(delimiter="  ")
      header = 'Valid:'
      batch_iterator = metric_logger.log_every(data_loader, 100, header)
  else:
      model = model.cuda()
      data_loader =  iter(data.DataLoader(dataset, args.batch_size, shuffle=False,
                                                num_workers=args.num_workers,
                                                collate_fn=coco_collate))
      batch_iterator = iter(data_loader)
```
`ModifiedFRCNN` is a class that inherits `FasterRCNN` to make trivial changes, such as parameter, postprocessing etc.
Case 1 : When n_gpu=1, I am able to use a batch size of upto 8.
Case 2 : When n_gpu=4, I am unable to even use a batch size of 1.

Both the above mentioned cases are on same the GPU, 2080Ti. 

## Expected behavior

Consume comparable memory if not equal on each GPUs as the case of training on a single GPU.

## Environment
```
Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Debian GNU/Linux 10 (buster)
GCC version: (Debian 8.3.0-6) 8.3.0
CMake version: version 3.13.4

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 430.14
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.19.0
[pip3] torch==1.2.0
[pip3] torchvision==0.4.0
```
## Additional context

The command I use to launch

```
python -m torch.distributed.launch --nproc_per_node=4 --use_env test.py <other_arguments> --world_size 4 --n_gpu 4
```

## PS

I have posted this issue [here](https://discuss.pytorch.org/t/excessive-memory-consumption-while-using-distributeddataparallel/87568) and since I did not receive any response, I was not sure whether the place where I posted this was correct, hence re-posting here. Apologies if that shouldn't be done.

Thank you!

**Comment by fmassa** [comment]
Hi,

The discussion forums is generally the best place to ask those questions, but let me quickly try to give one piece of feedback.

I would say that it's not expected to get a much larger memory consumption when using 4 GPUs compared to 1 GPU.

One thing to notice is that you seem to be overriding a number of flags which are generally not needed to be set in a standard reference script, like `world_size`, as they are handled by `torch.distributed.launch` and the rest of the reference scripts.

I would recommend you taking the reference scripts from https://github.com/pytorch/vision/tree/master/references/detection, they are the scripts we use to train the models in torchvision and are meant to be fairly simple.
Running those scripts on a single GPU can be done via 
```
python -m torch.distributed.launch --nproc_per_node=1 --use_env train.py\
    --dataset coco --model fasterrcnn_resnet50_fpn --epochs 26\
    --lr-steps 16 22 --aspect-ratio-group-factor 3
```
and then on 4 GPUs via
```
python -m torch.distributed.launch --nproc_per_node=4 --use_env train.py\
    --dataset coco --model fasterrcnn_resnet50_fpn --epochs 26\
    --lr-steps 16 22 --aspect-ratio-group-factor 3
```
(note that the only change was the `nproc_per_node`). This should give you an idea on what to expect wrt the memory, and it will probably help you figure out what the issue is in your codebase.

Best of luck!
